File: src/main/java/dev/psygamer/wireframe/util/BlockPosition.java
package dev.psygamer.wireframe.util;

import dev.psygamer.wireframe.util.math.vector.Vector3i;

public class BlockPosition extends Vector3i {
	
	public BlockPosition(final int x, final int y, final int z) {
		super(x, y, z);
	}
	
	public static BlockPosition get(final net.minecraft.util.math.BlockPos internalPosition) {
		return new BlockPosition(internalPosition.getX(), internalPosition.getY(), internalPosition.getZ());
	}
	
	public BlockPosition offset(final Direction direction) {
		return offset(1, direction);
	}
	
	public BlockPosition offset(final int distance, final Direction direction) {
		return offset(
				direction.getUnitVector().getX() * distance,
				direction.getUnitVector().getY() * distance,
				direction.getUnitVector().getZ() * distance
		);
	}
	
	public BlockPosition offset(final int xOffset, final int yOffset, final int zOffset) {
		return new BlockPosition(
				getX() + xOffset,
				getY() + yOffset,
				getZ() + zOffset
		);
	}
	
	public net.minecraft.util.math.BlockPos toInternal() {
		return new net.minecraft.util.math.BlockPos(this.getX(), this.getY(), this.getZ());
	}
}

File: src/main/kotlin/dev/psygamer/wireframe/util/BlockPosition.kt
package dev.psygamer.wireframe.util

import org.joml.Vector3i

class BlockPosition(x: Int, y: Int, z: Int) : Vector3i(x, y, z) {

	@JvmOverloads
	fun offset(direction: Direction, distance: Int = 1): BlockPosition {
		return offset(
			direction.unitVector.x * distance,
			direction.unitVector.y * distance,
			direction.unitVector.z * distance
		)
	}

	fun offset(xOffset: Int, yOffset: Int, zOffset: Int): BlockPosition {
		return BlockPosition(
			x + xOffset,
			y + yOffset,
			z + zOffset
		)
	}
}
